code/modules/client/client_procs.dm

```diff
@@ -836,15 +836,61 @@ var/list/localhost_addresses = list(
 	var/obj/O = object
 	if(istype(I, /obj/item/gun) && !mob.in_throw_mode)
 		var/obj/item/gun/G = I
-		if(G.can_autofire(O, location, params) && O.is_auto_clickable() && !(G.safety()) && !(G == O))
-			autofire_aiming_at[1] = O
-			autofire_aiming_at[2] = params
-			var/accuracy_dec = 0
-			while(autofire_aiming_at[1])
-				G.Fire(autofire_aiming_at[1], mob, autofire_aiming_at[2], (get_dist(mob, location) <= 1), FALSE, accuracy_dec)
-				mob.set_dir(get_dir(mob, autofire_aiming_at[1]))
-				accuracy_dec = min(accuracy_dec + 0.25, 2)
-				sleep(G.fire_delay)
+
+		// check if our gun can even autofire
+		var/can_autofire = G.can_autofire(O, location, params)
+		if(!can_autofire)
+			return
+
+		// check if the object we're clicking is auto clickable
+		var/can_autoclick_object = O.is_auto_clickable()
+		if(!can_autoclick_object)
+			return
+
+		// check if safety should block the shot
+		var/not_on_safety = mob.a_intent == I_HURT || (!G.safety())
+		if(!not_on_safety)
+			return
+
+		// check that we're not shooting anything on our person
+		var/object_not_on_us = O.loc != mob
+		if(!object_not_on_us)
+			return
+
+		// check whether we're putting it in a storage item / table / locker
+		var/object_not_storage_item = TRUE
+
+		var/is_adjacent_to_object = get_last_atom_before_turf(G) == mob
+		if(!is_adjacent_to_object)
+			var/distance_to_object = get_dist(mob, location)
+			is_adjacent_to_object = distance_to_object <= 1
+
+		if(is_adjacent_to_object) // if we're not next to it, blast it
+			var/list/storage_types = list(
+				/obj/item/storage,
+				/obj/structure/table,
+				/obj/structure/closet
+			)
+			var/is_storage_item = is_type_in_list(O, storage_types)
+			if(!is_storage_item)
+				is_storage_item = locate(/obj/item/storage) in O // some items will have storage within themselves
+			object_not_storage_item = !is_storage_item
+
+		if(!object_not_storage_item)
+			return
+
+		autofire_aiming_at[1] = O
+		autofire_aiming_at[2] = params
+
+		var/accuracy_dec = 0
+		while(autofire_aiming_at[1])
+			// stop shooting if we drop our gun
+			if(G.loc != mob)
+				break
+			G.Fire(autofire_aiming_at[1], mob, autofire_aiming_at[2], (get_dist(mob, location) <= 1), FALSE, accuracy_dec)
+			mob.set_dir(get_dir(mob, autofire_aiming_at[1]))
+			accuracy_dec = min(accuracy_dec + 0.25, 2)
+			sleep(G.fire_delay)
 			CHECK_TICK
```

Comment:
Yelding to CHECK_TICK from a proc that's meant to handle user inputs seems like a terrible idea, I understand however that you didn't wrote this, but sinde there's a sleep, this CHECK_TICK seems between bad and pointless, even sleep(0) would yeld until the end of the tick?

Reply:
i'm not sure tbh, i'm not too in the loop with how ticks and stuff works, so if you have a recommendation, lemme know

Reply:
I'm 98% sure it shouldn't be there and can / is better to remove it; comparing with other codebases, neither TG nor Bay have it either